python/ray/serve/tests/test_util.py

```diff
@@ -142,6 +143,109 @@ def test_replace_main(self):
             subprocess.check_output(["python", full_fname])
 
 
+class TestMergeRuntimeEnvs:
+    def test_merge_empty(self):
+        assert {"env_vars": {}} == merge_runtime_envs({}, {})
+
+    def test_merge_empty_parent(self):
+        child = {"env_vars": {"test1": "test_val"}, "working_dir": "."}
+        assert child == merge_runtime_envs({}, child)
+
+    def test_merge_empty_child(self):
+        parent = {"env_vars": {"test1": "test_val"}, "working_dir": "."}
+        assert parent == merge_runtime_envs(parent, {})
+
+    @pytest.mark.parametrize("invalid_env", [None, 0, "runtime_env", set()])
+    def test_invalid_type(self, invalid_env):
+        with pytest.raises(TypeError):
+            merge_runtime_envs(invalid_env, {})
+        with pytest.raises(TypeError):
+            merge_runtime_envs({}, invalid_env)
+        with pytest.raises(TypeError):
+            merge_runtime_envs(invalid_env, invalid_env)
+
+    def test_basic_merge(self):
+        parent = {
+            "py_modules": ["http://test.com/test0.zip", "s3://path/test1.zip"],
+            "working_dir": "gs://path/test2.zip",
+            "env_vars": {"test": "val", "trial": "val2"},
+            "pip": ["pandas", "numpy"],
+            "excludes": ["my_file.txt"],
+        }
+        original_parent = parent.copy()
+
+        child = {
+            "py_modules": [],
+            "working_dir": "s3://path/test1.zip",
+            "env_vars": {"test": "val", "trial": "val2"},
+            "pip": ["numpy"],
+        }
+        original_child = child.copy()
+
+        merged = merge_runtime_envs(parent, child)
+
+        assert original_parent == parent
+        assert original_child == child
+        assert merged == {
+            "py_modules": [],
```

Comment:
wait py_modules isn't append or prepend?

Reply:
i'm a bit missing the context again, can you remind me why are we doing override here?

Reply:
Sure, this override logic matches the current [`runtime_env` inheritance rules](https://docs.ray.io/en/releases-1.12.1/ray-core/handling-dependencies.html#inheritance). From that documentation:

> Every other field [than `env_vars`] in the `runtime_env` will be overridden by the child, not merged. For example, if `runtime_env["py_modules"]` is specified, it will replace the `runtime_env["py_modules"]` field of the parent.